Removed infos vector in torch.linalg.qr

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56257 Fix MAGMA qr for empty batched inputs
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* #56254 Add non-allocating helper function for torch.linalg.qr
* #56253 Add cuBLAS path for batched torch.geqrf
* #56252 Add cuSOLVER path for torch.geqrf
* #56251 Port CUDA torch.geqrf to ATen
* #56250 Update internal code for torch.geqrf
* #56249 Port CPU torch.geqrf to ATen
* **#56248 Removed infos vector in torch.linalg.qr**

`info` error code for QR decomposition only indicates wrong parameters,
when everything is implemented correctly it will never be nonzero so we
don't need to check it for CPU path.
For MAGMA `checkMagmaInternalError` is added that checks for failed
memory allocations internal to MAGMA.

Differential Revision: [D27850414](https://our.internmc.facebook.com/intern/diff/D27850414)

IvanYashchuk committed:
Removed infos vector in torch.linalg.qr

`info` error code for QR decomposition only indicates wrong parameters,
when everything is implemented correctly it will never be nonzero so we
don't need to check it for CPU path.
For MAGMA `checkMagmaInternalError` is added that checks for failed
memory allocations internal to MAGMA.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1192,7 +1192,40 @@ void magmaGels<c10::complex<double>>(
+*/
+void checkMagmaInternalError(magma_int_t info, const std::string& magma_function_name) {
+  // if info > 0 the error is function-specific, do nothing in this case
+  TORCH_CHECK(info >= 0,
```
What's an example for the value of MAGMA_ERR_NOT_IMPLEMENTED?

Also won't this check trigger both when the ith argument has an illegal value and when a MAGMA-specific error code is returned?

IvanYashchuk replied:
I think we don't use any function from MAGMA that could give MAGMA_ERR_NOT_IMPLEMENTED.

Yes, this check would trigger both when the ith argument has an illegal value and when a MAGMA-specific error code is returned. If two problems are present what is actually returned depends on magma's source code. This error check doesn't cause device synchronization because the `magma_int_t info` is always on host (CPU) memory.

IvanYashchuk replied:
`magma_strerror` gives a string both for internal failures and invalid arguments.
See https://bitbucket.org/icl/magma/src/4c2588119ee636999b230b248fca3e3824a23be9/interface_cuda/error.cpp?at=master#lines-137

IvanYashchuk replied:
I added printing the actual value of `info` to the error message so that we know which argument to the call is invalid. For other error types, the value is meaningless.

mruberry reviewed (approved):
Cool!

IvanYashchuk committed:
Update on "Removed infos vector in torch.linalg.qr"


`info` error code for QR decomposition only indicates wrong parameters,
when everything is implemented correctly it will never be nonzero so we
don't need to check it for CPU path.
For MAGMA `checkMagmaInternalError` is added that checks for failed
memory allocations internal to MAGMA.

Differential Revision: [D27850414](https://our.internmc.facebook.com/intern/diff/D27850414)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Removed infos vector in torch.linalg.qr"


`info` error code for QR decomposition only indicates wrong parameters,
when everything is implemented correctly it will never be nonzero so we
don't need to check it for CPU path.
For MAGMA `checkMagmaInternalError` is added that checks for failed
memory allocations internal to MAGMA.

Differential Revision: [D27850414](https://our.internmc.facebook.com/intern/diff/D27850414)

[ghstack-poisoned]
